generated_code/DynamicRupture.py

```diff
@@ -95,4 +97,39 @@ def interpolateQGenerator(i,h):
                         nodalFluxPrefetch if target =='cpu' else None,
                         target=target)
 
+  # Energy output
+  # Minus and plus refer to the original implementation of Christian Pelties,
+  # where the normal points from the plus side to the minus side
+  QInterpolatedPlus = OptionalDimTensor('QInterpolatedPlus', aderdg.Q.optName(), aderdg.Q.optSize(), aderdg.Q.optPos(), gShape, alignStride=True)
+  QInterpolatedMinus = OptionalDimTensor('QInterpolatedMinus', aderdg.Q.optName(), aderdg.Q.optSize(), aderdg.Q.optPos(), gShape, alignStride=True)
+  slipRateInterpolated = Tensor('slipRateInterpolated', (numberOfPoints,3))
+  slipInterpolated = Tensor('slipInterpolated', (numberOfPoints,3))
+  absoluteSlipInterpolated = Tensor('absoluteSlipInterpolated', (numberOfPoints,))
+  tractionInterpolated = Tensor('tractionInterpolated', (numberOfPoints,3))
+  frictionalEnergy = Tensor('frictionalEnergy', ())
+  timeWeight = Scalar('timeWeight')
+  spaceWeights = Tensor('spaceWeights', (numberOfPoints,))
+
+  selectTractionSpp = np.zeros((aderdg.numberOfQuantities(), 3), dtype=bool)
+  selectTractionSpp[0,0] = True
+  selectTractionSpp[3,1] = True
+  selectTractionSpp[5,2] = True
+  tractionPlusMatrix = Tensor('tractionPlusMatrix', selectTractionSpp.shape, selectTractionSpp, CSCMemoryLayout)
+  tractionMinusMatrix = Tensor('tractionMinusMatrix', selectTractionSpp.shape, selectTractionSpp, CSCMemoryLayout)
+
+  computeSlipRateInterpolated = slipRateInterpolated['kp'] <= QInterpolatedMinus['kq'] * aderdg.selectVelocity['qp'] - QInterpolatedPlus['kq'] * aderdg.selectVelocity['qp']
+  generator.add('computeSlipRateInterpolated', computeSlipRateInterpolated)
+
+  computeTractionInterpolated = tractionInterpolated['kp'] <= QInterpolatedMinus['kq'] * tractionMinusMatrix['qp'] + QInterpolatedPlus['kq'] * tractionPlusMatrix['qp']
+  generator.add('computeTractionInterpolated', computeTractionInterpolated)
+
+  addSlipRateInterpolated = slipInterpolated['kp'] <= slipInterpolated['kp'] + timeWeight * slipRateInterpolated['kp']
+  generator.add('addSlipRateInterpolated', addSlipRateInterpolated)
+
+  squareSlipRateInterpolated = absoluteSlipInterpolated['k'] <= slipRateInterpolated['kp'] * slipRateInterpolated['kp']
+  generator.add('squareSlipRateInterpolated', squareSlipRateInterpolated)
```

Comment:
Doesn't this also sum over index p?

Reply:
Should I rename it to squaredNormSlipRateInterpolated?

Reply:
Yes, I think that's better

Reply:
@sebwolf-de yes you are right, this absoluteSlip is not used anywhere.  I just removed the kernel.